### PR TITLE
[SYCL] Do not use current directory for JIT cache

### PR DESCRIPTION
**Pull request by vladimirlaz**
Remove storing JIT cache files to current directory when corresponding
environment variables are not set.

Complementary test change: intel/llvm-test-suite#337

### CODE REVIEW

**Review by bader — CHANGES_REQUESTED** [review]
The documents update looks okay, but we need to add regression unit-tests.

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
I missed the comment about llvm-test-suite tests in a separate PR.

**Review by sergey-semenov — APPROVED** [review]

**Review by sergey-semenov — APPROVED** [review]

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 1]
```diff
@@ -91,6 +91,8 @@ class PersistenDeviceCodeCache : public ::testing::Test {
       return;
     }
 
+    set_env("SYCL_CACHE_DIR", ".");
+
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:95]: I don't think implicitly using current directory to store JIT cash for SYCL unittests is a good approach (especially if the test is launched manually). Is it possible to require explicit cache directory configuration or fail the test?

Comment (vladimirlaz) [reply sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:95]: The current directory in this case is <build_root>/tools/sycl/test/test_cache.
I think it is ok to have LIT test artifacts in the build directory.

Comment (bader) [reply sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:95]: > The current directory in this case is <build_root>/tools/sycl/test/test_cache.

This is wrong assumption in general. I think what you are trying to say is that:
CI system runs the test using `<build_root>/tools/sycl/test/test_cache` as current directory and clears `build_root` directory, so there should be no **additional** side effects from this change.

I agree with that, but that scenario was not my concern. I'm asking about other use cases e.g. developer manually run the test from non-`build_root` directory.

Comment (vladimirlaz) [reply sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp]: I have moved setting the environment variable to LIT infra and add the assertion to guide a user if he runs tests manually.

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 2]
```diff
@@ -81,6 +81,10 @@ class PersistenDeviceCodeCache : public ::testing::Test {
     return _putenv_s(name, value);
   }
 #endif
+  virtual void SetUp() {
+    assert(getenv("SYCL_CACHE_DIR") && "Please set SYCL_CACHE_DIR environment "
+                                       "variable pointing to cache location.");
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:86]: I think we should fail the test in Release configuration as well.

**Review by bader — COMMENTED** [review]
Much better, but Release configuration of the test still should be improved.

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp (hunk 1 above)
Comment (bader) [reply sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp]: Assertion is NOP in Release. I think we need the test to work in this configuration as well. Right?

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 3]
```diff
@@ -245,21 +251,23 @@ TEST_F(PersistenDeviceCodeCache, CorruptedCacheFiles) {
   // Only source file is present
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(!llvm::sys::fs::remove(ItemDir + "/0.bin") &&
-         "Failed to remove binary file");
+  EXPECT_EQ(!llvm::sys::fs::remove(ItemDir + "/0.bin"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:254]:
```suggestion
  EXPECT_FALSE(llvm::sys::fs::remove(ItemDir + "/0.bin"))
```
https://google.github.io/googletest/reference/assertions.html#boolean

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 4]
```diff
@@ -245,21 +251,23 @@ TEST_F(PersistenDeviceCodeCache, CorruptedCacheFiles) {
...
+  EXPECT_EQ(Res.size(), static_cast<size_t>(0))
+      << "Item with missed binary file was read";
   llvm::sys::fs::remove_directories(ItemDir);
 
   // Only binary file is present
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(!llvm::sys::fs::remove(ItemDir + "/0.src") &&
-         "Failed to remove source file");
+  EXPECT_EQ(!llvm::sys::fs::remove(ItemDir + "/0.src"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:265]:
```suggestion
  EXPECT_FALSE(llvm::sys::fs::remove(ItemDir + "/0.src"))
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 5]
```diff
@@ -272,10 +280,12 @@ TEST_F(PersistenDeviceCodeCache, CorruptedCacheFiles) {
    */
   FileStream << 2 << 12 << "123456789012" << 23 << "1234";
   FileStream.close();
-  assert((!FileStream.fail()) && "Failed to create trancated binary file");
+  EXPECT_EQ(FileStream.fail(), false)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:283]:
```suggestion
  EXPECT_FALSE(FileStream.fail())
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 6]
```diff
@@ -312,37 +323,40 @@ TEST_F(PersistenDeviceCodeCache, LockFile) {
   // Create 1st cahe item
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(llvm::sys::fs::exists(ItemDir + "/0.bin") && "No file created");
+  EXPECT_EQ(llvm::sys::fs::exists(ItemDir + "/0.bin"), true)
+      << "No file created";
   std::string LockFile = ItemDir + "/0.lock";
-  assert(!llvm::sys::fs::exists(LockFile) && "Cache item locked");
+  EXPECT_NE(llvm::sys::fs::exists(LockFile), true) << "Cache item locked";
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:329]:
```suggestion
  EXPECT_FALSE(llvm::sys::fs::exists(LockFile)) << "Cache item locked";
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 7]
```diff
@@ -312,37 +323,40 @@ TEST_F(PersistenDeviceCodeCache, LockFile) {
   // Create 1st cahe item
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(llvm::sys::fs::exists(ItemDir + "/0.bin") && "No file created");
+  EXPECT_EQ(llvm::sys::fs::exists(ItemDir + "/0.bin"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:326]:
```suggestion
  EXPECT_TRUE(llvm::sys::fs::exists(ItemDir + "/0.bin"))
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 8]
```diff
@@ -312,37 +323,40 @@ TEST_F(PersistenDeviceCodeCache, LockFile) {
...
   auto Res = detail::PersistentDeviceCodeCache::getItemFromDisc(Dev, Img, {},
                                                                 BuildOptions);
-  assert(Res.size() == 0 && "Locked item was read");
+  EXPECT_EQ(Res.size(), static_cast<size_t>(0)) << "Locked item was read";
 
   // Cache item is locked - new cache item to be created
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(llvm::sys::fs::exists(ItemDir + "/1.bin") && "No file created");
+  EXPECT_EQ(llvm::sys::fs::exists(ItemDir + "/1.bin"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:342]:
```suggestion
  EXPECT_TRUE(llvm::sys::fs::exists(ItemDir + "/1.bin"))
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 9]
```diff
@@ -366,19 +380,22 @@ TEST_F(PersistenDeviceCodeCache, AccessDeniedForCacheDir) {
   llvm::sys::fs::remove_directories(ItemDir);
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(llvm::sys::fs::exists(ItemDir + "/0.bin") && "No file created");
+  EXPECT_EQ(llvm::sys::fs::exists(ItemDir + "/0.bin"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:383]:
```suggestion
  EXPECT_TRUE(llvm::sys::fs::exists(ItemDir + "/0.bin"))
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 10]
```diff
@@ -366,19 +380,22 @@ TEST_F(PersistenDeviceCodeCache, AccessDeniedForCacheDir) {
...
+      << "No file created";
   llvm::sys::fs::setPermissions(ItemDir + "/0.bin", llvm::sys::fs::no_perms);
   // No access to binary file new cache item to be created
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  assert(llvm::sys::fs::exists(ItemDir + "/1.bin") && "No file created");
+  EXPECT_EQ(llvm::sys::fs::exists(ItemDir + "/1.bin"), true)
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:389]:
```suggestion
  EXPECT_TRUE(llvm::sys::fs::exists(ItemDir + "/1.bin"))
```

File: sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp [hunk 11]
```diff
@@ -262,7 +262,7 @@ TEST_F(PersistenDeviceCodeCache, CorruptedCacheFiles) {
   // Only binary file is present
   detail::PersistentDeviceCodeCache::putItemToDisc(Dev, Img, {}, BuildOptions,
                                                    NativeProg);
-  EXPECT_EQ(!llvm::sys::fs::remove(ItemDir + "/0.src"), true)
+  EXPECT_TRUE(!llvm::sys::fs::remove(ItemDir + "/0.src"))
```
Comment (bader) [inline sycl/unittests/kernel-and-program/PersistentDeviceCodeCache.cpp:265]:
```suggestion
  EXPECT_FALSE(llvm::sys::fs::remove(ItemDir + "/0.src"))
```

**Review by bader — APPROVED** [review]
Thanks!

**Review by sergey-semenov — APPROVED** [review]

**Comment by vladimirlaz** [comment]
@pvchupin could you please review/approve?

**Review by pvchupin — APPROVED** [review]

**Pull request merged**